pkg/driver/node.go

```diff
@@ -620,13 +620,41 @@ func (d *nodeService) nodePublishVolumeForFileSystem(req *csi.NodePublishVolumeR
 	mountOptions = collectMountOptions(fsType, mountOptions)
 
 	klog.V(4).Infof("NodePublishVolume: mounting %s at %s with option %s as fstype %s", source, target, mountOptions, fsType)
-	if err := d.mounter.Mount(source, target, fsType, mountOptions); err != nil {
-		if removeErr := os.Remove(target); removeErr != nil {
-			return status.Errorf(codes.Internal, "Could not remove mount target %q: %v", target, err)
+	/*
+		Checking if it's a mount point. There are three cases,
+		1. true, err when the directory does not exist or corrupted.
+		2. false, nil when the path is already mounted with a device.
+		3. true, nil when the path is not mounted with any device.
+	*/
+	notMnt, err := d.mounter.IsLikelyNotMountPoint(target)
```

Comment:
An easy way to cover block devices as well is to move the check right below the in flight check.

You may want to take a look at the solutions both [GCP PD](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/a26ef36a9e6bf92cc32ee93bb72b2a985d61e3fe/pkg/gce-pd-csi-driver/node.go#L117) and Azure Disk implemented.

Reply:
@bertinatto Thanks. New commit covering block devices is pushed.